File: src/main/java/javaclasses/Strings/LongestSubstringWithoutRepeating.java
package javaclasses.Strings;
/*Given a string, find the length of the longest substring without repeating characters.

For example –

Example 1:

Input: “abcabcbb” Output: 3

The output string is “abc”, with a length of 3.

Example 2:

Input: “bbbbb” Output: 1

The longest substring in this example is “b”. Its length is 1.

Example 3:

Input: “pwwkew” Output: 3

The answer is “wke”. Its length is 3.

*/

/*In this example, I am going to explain how to find the length of longest non-repeating substring using sliding window approach.


The idea here is to maintain a window of unique character. Each window has start and end index, based on that we know the window size.
To check unique character in a window, i am using set data structure. Set does not allow duplicate values and also the lookup time is O(1).

Here are the following steps to solve this problem –

Declare two indexes i and j and initialize with zero. Variable i indicate index of the start window and j indicate end index.
Traverse a string and pick one character at a time.
First check, if a character exists in a set. If it doesn’t exist in a set then add them in a set and increment the count of j and the index of i remain as it is.
Also, keep track of the length of a window.
If character exists in a set then remove the character from a set and increment the count of i until all the characters in a window is unique again.
Repeat this step until the string is traversed completely.
The time complexity of this approach is O(n) and it’s space complexity is also O(n).*/


/*
Java code to find the length of longest substring 
without repeating characters.
*/

import java.util.HashSet;
import java.util.Set;

public class LongestSubstringWithoutRepeating {

    public static int lengthOfLongestSubstring(String s) {

        int maxCount = 0;
        int i = 0;
        int j = 0;
        int strLen = s.length();

        //Declare set of characters
        Set<Character> st = new HashSet<>();

        //Traverse a string
        while (i < strLen && j < strLen) {

            //If it's a unique character add in a set
            if (!st.contains(s.charAt(j))) {
                st.add(s.charAt(j));
                j++;
                maxCount = Math.max(maxCount, j - i);
            } else {
                st.remove(s.charAt(i));
                i++;
            }
        }

        return maxCount;
    }

    public static void main(String[] args) {

        String str = "abcdabcdbb";
        int len = lengthOfLongestSubstring(str);

        System.out.println(" Length " + len);
    }
}

File: src/main/java/kotlinclasses/TwoPointer_SlidingWindow/LongestSubstringWithoutRepeating.kt
/*
 * *
 *  * Longest Substring Without Repeating Characters.kt
 *  * Created by Rafsan Ahmad on 6/24/25, 8:08PM
 *  * Copyright (c) 2025 . All rights reserved.
 *  *
 *
 */

package kotlinclasses.TwoPointer_SlidingWindow

import kotlin.math.max

class LongestSubstringWithoutRepeating {
    //https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
    /*Given a string s, find the length of the longest substring without duplicate characters.

Example 1:
Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.

Example 2:
Input: s = "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.

Example 3:
Input: s = "pwwkew"
Output: 3
Explanation: The answer is "wke", with the length of 3.
Notice that the answer must be a substring, "pwke" is a subsequence and not a substring.

Constraints:
0 <= s.length <= 5 * 10^4
s consists of English letters, digits, symbols and spaces.*/

    fun lengthOfLongestSubstring(s: String): Int {
        var maxCount = 0
        var i = 0
        var j = 0
        val strLen = s.length

        //Declare set of characters
        val st: MutableSet<Char> = hashSetOf()

        //Traverse a string
        while (i < strLen && j < strLen) {
            //If it's a unique character add in a set

            if (!st.contains(s[j])) {
                st.add(s[j])
                j++
                maxCount = max(maxCount, (j - i))
            } else {
                st.remove(s[i])
                i++
            }
        }

        return maxCount
    }

    fun lengthOfLongestSubstring2(s: String): Int {
        val map = mutableMapOf<Char, Int>()
        var result = 0
        var left = 0

        for (right in s.indices) {
            val ch = s[right]
            val prevIndex = map[ch]
            if (prevIndex != null) {
                left = maxOf(left, prevIndex + 1)
            }
            map[ch] = right
            result = maxOf(result, right - left + 1)
        }

        return result
    }
}

fun main() {
    val obj = LongestSubstringWithoutRepeating()
    println(obj.lengthOfLongestSubstring("abcabcbb"))
    println(obj.lengthOfLongestSubstring("pwwkew"))
    println(obj.lengthOfLongestSubstring2("abcabcbb"))
    println(obj.lengthOfLongestSubstring2("pwwkew"))
}